### PR TITLE
Electrodialysis 0 d

### PR DESCRIPTION
**Pull request by lbibl**
## Summary/Motivation:
- This branch adds the first version of Electrodialysis_0D unit model, including the main model file and test file. 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/unit_models/electrodialysis_0d.py [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+        # Create essential sets.
+        self.membrane_set = Set(initialize=["cem", "aem"])
+
+        # Create unit model parameters and vars
+        self.water_density = Param(
+            initialize=1000,
+            mutable=False,
+            units=pyunits.kg * pyunits.m**-3,
+            doc="density of water",
+        )
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py:204]: Remind me again, why is this here? This should be a property calculated in the property package. What is this used for and why can't we use the property package?

File: watertap/unit_models/electrodialysis_0d.py [hunk 2]
```diff
@@ -0,0 +1,938 @@
...
+        self.water_MW = Param(
+            initialize=18.015e-3,
+            mutable=False,
+            units=pyunits.kg * pyunits.mole**-1,
+            doc="molecular weight of water",
+        )
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]: This is a piece of information that should be coming from the property package. Why do we need this here?

File: watertap/unit_models/electrodialysis_0d.py [hunk 3]
```diff
@@ -0,0 +1,938 @@
...
+        self.cell_pair_num = Var(
+            initialize=1,
+            domain=NonNegativeIntegers,
+            bounds=(1, Inf),
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
            bounds=(1, 1000),
```
We should actually bound this variable to something reasonable.

File: watertap/unit_models/electrodialysis_0d.py [hunk 4]
```diff
@@ -0,0 +1,938 @@
...
+            bounds=(1e-6, 1e-1),
+            units=pyunits.meter,
+            doc="Membrane thickness",
+        )
+        self.solute_diffusivity_membrane = Var(
+            self.membrane_set,
+            self.config.property_package.ion_set
+            | self.config.property_package.solute_set,
+            initialize=1e-10,
+            # bounds=(1e-11, 1e-6),
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
            bounds=(1e-11, 1e-6),
```
Why is this being commented out? Is the lower bound low enough for this variable?

File: watertap/unit_models/electrodialysis_0d.py [hunk 5]
```diff
@@ -0,0 +1,938 @@
...
+        self.ion_trans_number_membrane = Var(
+            self.membrane_set,
+            self.config.property_package.ion_set,
+            bounds=(0, 1),
+            units=pyunits.dimensionless,
+            doc="Ion transference number in the membrane phase",
+        )
+        self.water_trans_number_membrane = Var(
+            self.membrane_set,
+            initialize=5,
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py:259]:
```suggestion
            initialize=1,
```
Don't initialize a value outside of its bounds.

File: watertap/unit_models/electrodialysis_0d.py [hunk 6]
```diff
@@ -0,0 +1,938 @@
...
+            bounds=(0, 1),
+            units=pyunits.dimensionless,
+            doc="Transference number of water in membranes",
+        )
+        self.water_permeability_membrane = Var(
+            self.membrane_set,
+            initialize=5,
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
            initialize=1e-14,
```
We should not be giving an unreasonable default value for any variable.

File: watertap/unit_models/electrodialysis_0d.py [hunk 7]
```diff
@@ -0,0 +1,938 @@
...
+            self.flowsheet().config.time,
+            initialize=0.9,
+            bounds=(0, 1),
+            units=pyunits.dimensionless,
+            doc="The overall current efficiency for deionizaiton",
+        )
+        self.power_electrical = Var(
+            self.flowsheet().config.time,
+            initialize=1,
+            bounds=(0, Inf),
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
            bounds=(0, 1e5),
```
We should try and provide a reasonable upper limit (whatever that limit may be)

File: watertap/unit_models/electrodialysis_0d.py [hunk 8]
```diff
@@ -0,0 +1,938 @@
...
+            domain=NonNegativeReals,
+            units=pyunits.watt,
+            doc="Electrical power consumption of a stack",
+        )
+        self.specific_power_electrical = Var(
+            self.flowsheet().config.time,
+            initialize=1e6,
+            bounds=(0, Inf),
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
            bounds=(0, 1e5),
```
Provide reasonable upper limits to variables.

File: watertap/unit_models/electrodialysis_0d.py [hunk 9]
```diff
@@ -0,0 +1,938 @@
...
+    **default** - False. The filtration unit does not have defined volume, thus
+    this must be False.""",
+        ),
+    )
+
+    CONFIG.declare(
+        "operation_mode",
+        ConfigValue(
+            default="Constant Current",
+            domain=In(["Constant Current", "Constant Voltage"]),
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
            default="const_current",
            domain=In(["const_current", "const_voltage"]),
```
Avoid spaces in names of options, variables, and objects. It may also be better if these were done as an Enum, rather than strings.

File: watertap/unit_models/electrodialysis_0d.py [hunk 10]
```diff
@@ -0,0 +1,938 @@
...
+            )
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            doc="Equation for electrical migration flux_in",
+        )
+        def eq_elec_migration_flux_in(self, t, p, j):
+            if j == "H2O":
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py:448]: Because the unit model specifically requires a component named 'H2O', we may want to check the property package component list during the build of the object and raise a configuration error if 'H2O' is not present. This way, the user will get a clear message from the unit model saying why things have failed and telling them they need 'H2O' as a component in their property package.

File: watertap/unit_models/electrodialysis_0d.py [hunk 11]
```diff
@@ -0,0 +1,938 @@
...
+                        "Liq", j
+                    ]
+                    * self.config.property_package.charge_comp[j]
+                    for j in self.config.property_package.cation_set
+                )
+                * Constants.faraday_constant
+            )
+
+    # initialize method
+    def initialize(
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
    def initialize_build(
```
I believe newer version of IDAES will require the `initialize` function for custom unit models to be called `initialize_build`. It is basically the same function, but has some different args. Have a look at #516 for an example.

File: watertap/unit_models/electrodialysis_0d.py [hunk 12]
```diff
@@ -0,0 +1,938 @@
...
+            sf = (
+                iscale.get_scaling_factor(self.solute_diffusivity_membrane)
+                / iscale.get_scaling_factor(self.membrane_thickness)
+                * iscale.get_scaling_factor(
+                    self.concentrate_channel.properties_in[ind[0]].conc_mol_phase_comp[
+                        ind[1], ind[2]
+                    ]
+                )
+            )
+            # print('########', ind, sf)
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py]:
```suggestion
```
Don't need this line anymore.

**Review by aladshaw3 — CHANGES_REQUESTED** [review]
Looking good so far. Some additional changes suggested below.

File: watertap/unit_models/tests/test_electrodialysis_0d.py [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+        m.fs.unit.config.operation_mode = "Constant Voltage"
+        assert len(m.fs.unit.config) == 7
+        assert not m.fs.unit.config.dynamic
+        assert not m.fs.unit.config.has_holdup
+        assert m.fs.unit.config.operation_mode == "Constant Voltage"
+        assert m.fs.unit.config.material_balance_type == MaterialBalanceType.useDefault
+        assert (
+            m.fs.unit.config.momentum_balance_type == MomentumBalanceType.pressureTotal
+        )
+        assert m.fs.unit.config.property_package is m.fs.properties
```
Comment (aladshaw3) [inline watertap/unit_models/tests/test_electrodialysis_0d.py]: These asserts should be in the `test_build_model` function below.

File: watertap/unit_models/tests/test_electrodialysis_0d.py [hunk 14]
```diff
@@ -0,0 +1,672 @@
...
+        m.fs.unit.config.operation_mode = "Constant Current"
+        assert len(m.fs.unit.config) == 7
+        assert not m.fs.unit.config.dynamic
+        assert not m.fs.unit.config.has_holdup
+        assert m.fs.unit.config.operation_mode == "Constant Current"
+        assert m.fs.unit.config.material_balance_type == MaterialBalanceType.useDefault
+        assert (
+            m.fs.unit.config.momentum_balance_type == MomentumBalanceType.pressureTotal
+        )
+        assert m.fs.unit.config.property_package is m.fs.properties
```
Comment (aladshaw3) [inline watertap/unit_models/tests/test_electrodialysis_0d.py]: These assertions should be in the `test_build_model` function below

File: watertap/unit_models/tests/test_electrodialysis_0d.py [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
```
Comment (aladshaw3) [inline watertap/unit_models/tests/test_electrodialysis_0d.py]: These assertions should be in the `test_build_model` function below

File: watertap/unit_models/tests/test_electrodialysis_0d.py [hunk 16]
```diff
@@ -0,0 +1,672 @@
...
+# through Lawrence Berkeley National Laboratory, Oak Ridge National
+# Laboratory, National Renewable Energy Laboratory, and National Energy
+# Technology Laboratory (subject to receipt of any required approvals from
+# the U.S. Dept. of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/nawi-hub/proteuslib/"
+#
+###############################################################################
```
Comment (aladshaw3) [inline watertap/unit_models/tests/test_electrodialysis_0d.py]: This is the old copyright statement from ProteusLib. This needs to be replaced with the copyright from WaterTAP (the one you used in the unit model). If this was copy-pasted from another file, then it needs to be changed in that other file as well. 

If you could, also replace the copyright statement in the file you got this from in this PR.

File: watertap/unit_models/electrodialysis_0d.py (hunk 1 above)
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0d.py:204]: This was left in the unit model because it was used to build a unit_model var that does not distinguish the two control_volumes (concentrate and diluate). Taking it from property pack  will have to bind it to `concentrate_channel` or `diluate_channel`. Although that'll be technically the same number, it may make the equation confusing.  So I would prefer to keep it here.

File: watertap/unit_models/electrodialysis_0d.py (hunk 2 above)
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0d.py]: I removed this one in the new commit and take the value from property pack.  No issue for this one because mw_comp is a universal param in the ion pp.

File: watertap/unit_models/electrodialysis_0d.py (hunk 9 above)
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0d.py]: I kept the strings but removed the space.   Maybe revisit the option of Enum if this config becomes more useful (for now it is more or less a "label" only).

File: watertap/unit_models/tests/test_electrodialysis_0d.py (hunk 16 above)
Comment (lbibl) [reply watertap/unit_models/tests/test_electrodialysis_0d.py]: Yes, I also corrected the file "test_coag_..."  where I took it from .

File: watertap/unit_models/electrodialysis_0d.py (hunk 10 above)
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0d.py:448]: I believe "H2O" is built as solvent always in the ion property pack. But I did add a `assert` for testing "H2O" in `component_list` in case other property packs are used.

File: watertap/unit_models/electrodialysis_0d.py (hunk 4 above)
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0d.py]: I put it back.  It should be a temporary disabling during my debug.  It does not cause a problem till now.

**Comment by lbibl** [comment]
New commits are pushed to address Austin's comments (Thank you for catching some small but important issues!)

File: watertap/unit_models/electrodialysis_0d.py (hunk 1 above)
Comment (aladshaw3) [reply watertap/unit_models/electrodialysis_0d.py:204]: Should we consider adding a temperature dependence to this term as part of a later model refinement?

File: watertap/unit_models/electrodialysis_0d.py [hunk 17]
```diff
@@ -0,0 +1,934 @@
...
+        self.ion_trans_number_membrane = Var(
+            self.membrane_set,
+            self.config.property_package.ion_set,
+            bounds=(0, 1),
+            units=pyunits.dimensionless,
+            doc="Ion transference number in the membrane phase",
+        )
+        self.water_trans_number_membrane = Var(
+            self.membrane_set,
+            initialize=5,
```
Comment (aladshaw3) [inline watertap/unit_models/electrodialysis_0d.py:259]: Is this correct? I thought these we numbers between 0 and 1?

File: watertap/unit_models/electrodialysis_0d.py (hunk 10 above)
Comment (aladshaw3) [reply watertap/unit_models/electrodialysis_0d.py:448]: What I meant was to check for 'H2O' in the build function of this unit model and raise an exception if not present (not to just check in the tests). This is because ideally, we may want to use other property packages with this unit model (if available). 

One thing to understand is that since property packages and unit models are separate entities, we may not always be able to assume what the structure of the property package is. Something to keep in mind for future models and refinement.

**Review by aladshaw3 — DISMISSED** [review]
I had a few other minor comments/responses. Other than that, this looks good to me.

File: watertap/unit_models/electrodialysis_0d.py (hunk 1 above)
Comment (lbibl) [reply watertap/unit_models/electrodialysis_0d.py:204]: I do not think it is technically necessary because water density variation with temperature from 0 to 60 Celsius degree is <2%, which would likely be well below the error of temperature measurements in reality.  The error caused by this would be even much less significant for the ED Perfromance modeling.   I would recommend leaving this out for now but we may revisit this if the model refinement shows this becomes a critical factor to change.

**Review by srikanthallu — COMMENTED** [review]
Can you provide references from where you got the values used in the testing of the unit_model ?

**Comment by lbibl** [comment]
> Can you provide references from where you got the values used in the testing of the unit_model ?

I put a brief note on the testing scenario reference  in the testing code file.  Formal references will be given in the technical documentation.

**Review by aladshaw3 — APPROVED** [review]

**Review by srikanthallu — APPROVED** [review]
